src/plugins/data/public/data_sources/datasource_selector/datasource_selectable.tsx

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import React, { useEffect } from 'react';
+import { i18n } from '@osd/i18n';
+import { EuiComboBox } from '@elastic/eui';
+import { DataSourceType, GenericDataSource } from '../datasource_services';
+import { DataSourceGroup, DataSourceSelectableProps } from './types';
+import { ISourceDataSet, IndexPatternOption } from '../datasource';
+
+type DataSourceTypeKey = 'DEFAULT_INDEX_PATTERNS' | 's3glue' | 'spark';
+
+// Mapping between datasource type and its display name.
+const DATASOURCE_TYPE_DISPLAY_NAME_MAP: Record<DataSourceTypeKey, string> = {
+  DEFAULT_INDEX_PATTERNS: 'Index patterns',
+  s3glue: 'Amazon S3',
+  spark: 'Spark',
+};
+
+type DataSetType = string | IndexPatternOption;
+
+interface DataSetWithSource {
+  ds: GenericDataSource;
+  data_sets: string[] | IndexPatternOption[];
+}
+
+export const DataSourceSelectable = ({
+  dataSources, // list of all available datasource connections.
+  dataSourceOptionList, // combo box renderable option list derived from dataSources
+  selectedSources, // current selected datasource in the form of [{ label: xxx, value: xxx }]
+  onDataSourceSelect,
+  setDataSourceOptionList,
+  onFetchDataSetError, //   onFetchDataSetError, Callback for handling dataset fetch errors. Ensure it's memoized.
+  singleSelection = true,
+}: DataSourceSelectableProps) => {
+  // This effect fetches datasets and prepares the datasource list for UI rendering.
+  useEffect(() => {
+    // Fetches datasets for a given datasource and returns it along with the source.
+    const fetchDataSetWithSource = async (ds: GenericDataSource): Promise<DataSetWithSource> => {
+      const dataSet = await ds.getDataSet();
+      return {
+        ds,
+        data_sets: dataSet,
+      };
+    };
+
+    // Map through all data sources and fetch their respective datasets.
+    const fetchDataSets = () => dataSources.map((ds: DataSourceType) => fetchDataSetWithSource(ds));
+
+    const isIndexPatterns = (dataset: string | IndexPatternOption): boolean => {
+      if (typeof dataset === 'string') return false;
+
+      return dataset.title !== undefined && dataset.id !== undefined;
```

Comment:
Does this mean we are okay with:
```js
dataset = { title: '', id: '' }
```

Reply:
Also, isn't `dataset` is defined with `DataSetType = any`. The guards here need to check against a lot more that `string`.

Reply:
Based on current type it is an index pattern if `title` and `id` exists on `dataSet`. The improvement should be in  https://github.com/opensearch-project/OpenSearch-Dashboards/blob/e8f219b3febaaa5d5d7ce21cd9c39dc560268e0e/src/plugins/data/public/data_sources/datasource/datasource.ts#L18-L24

and make `DataSource` to enforce `SourceDataSet` passed in to not be `any`

i checked with eric, the data set shouldn't be an index pattern if title or id is an empty string. I didn't understand the logic but will make the change